evennia/contrib/pathfinding/pathfinder.py

```diff
@@ -0,0 +1,394 @@
+as they grow.
+
+Using this class directly is really only appropriate for testing and experimenta-
+tion. For general usage, you'll want to use the helper Script included to create
+the "map" and keep it updated.
+
+    from evennia.contrib.pathfinding.pathfinder import Pathfinder
+
+    # Instantiation
+    pathfinder = Pathfinder()
+    
+    # Refresh the Pathfinder to add/remove new/old rooms and exits
+    pathfinder.update()
+    
+    # Get directions from one character to another
+    room1 = char1.location
+    room2 = char2.location
+    pathfinder.get_directions(room1, room2)
+    >> ['down', 'east', 'down', 'north', 'down']
+    
+    # Get line of sight between consecutive rooms, no further than 3 rooms away
+    # 'look east'
+    pathfinder.get_line_of_sight(room1, 'east', 3)
+    >> [<DefaultRoom: Railcar #0>, <DefaultRoom: Railcar #1>, <DefaultRoom: Railcar #2>]
+
+"""
+
+from evennia import DefaultExit as Exit, DefaultRoom as Room
+from evennia.locks.lockhandler import LockHandler
+from itertools import chain
+from networkx.readwrite import json_graph
+import networkx as nx
+import json
+
+# Define a dummy class to do lock checks against
+class Edge(object):
+    def __init__(self, edge, source_graph):
+        self.src = edge[0]
+        self.dst = edge[1]
+        
+        attrs = source_graph[self.src][self.dst]
+        self.key = attrs.get('key', attrs.get('direction'))
+        self.lock_storage = attrs['locks_str']
+        self.locks = LockHandler(self)
+
+class Pathfinder(nx.DiGraph):
+    """
+    Creates a directional graph of your entire network of rooms and exits.
+    """
+    # Arbitrary delimiter used to create unique keys to use as nodes.
+    # Should not be set it to anything that could conceivably be part of a room
+    # name since it will interfere with parsing.
+    delimiter = '[(:@_@:)]'
+    
+    def __init__(self, *args, **kwargs):
+        """
+        Creates the graph and initializes it.
+        
+        Kwargs:
+            queryset (QuerySet, optional): QuerySet of Room objects to be 
+                included in the graph
+                
+        """
+        self.queryset = kwargs.pop('queryset', None)
+        super(Pathfinder, self).__init__(*args, **kwargs)
+        
+    def get_key(self, room):
+        """
+        Creates unique keys for use as nodes in the graph.
+
+        Args:
+            room (Room): Room object needing a key.
+            
+        Returns:
+            key (str): Unique key identifying the node.
+        
+        """
+        return self.delimiter.join([room.db_key, str(room.id)])
+        
+    def get_queryset(self):
+        """
+        Get all the Room objects to be included in this graph.
+        
+        Returns:
+            objs (generator): Any iterable generator of Room objects you
+                wish to be part of the graph.
+            
+        """
+        if self.queryset: 
+            return (x for x in self.queryset)
```

Comment:
Since you call the variable `self.queryset`, it's not clear to the user that slapped-together lists would be ok. But sure, if you want to support that then this solution would work. You should probably make that clearer in the documentation in that case.